sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -76,68 +94,146 @@ public class BlobDecryptionPolicy implements HttpPipelinePolicy {
+                    .flatMap(encryptionData -> {
+                        EncryptedBlobRange encryptedRange = EncryptedBlobRange.getEncryptedBlobRangeFromHeader(
+                            initialRangeHeader, encryptionData);
+                        if (context.getHttpRequest().getHeaders().getValue(RANGE_HEADER) != null
+                            && encryptionData != null) {
+                            requestHeaders.set(RANGE_HEADER, encryptedRange.toBlobRange().toString());
+                        }
+                        return Mono.zip(Mono.just(encryptionData), Mono.just(encryptedRange));
+                    })
+                    .flatMap(tuple2 -> next.process().flatMap(httpResponse -> {
+                        if (httpResponse.getRequest().getHttpMethod() == HttpMethod.GET && httpResponse.getBody()
+                            != null) {
+                            HttpHeaders responseHeaders = httpResponse.getHeaders();
+                            // Checking that encryption data at least exists on the download call even if we didn't use
+                            // it for deserialization ensures that the download response was not an error response.
+                            if (httpResponse.getHeaderValue(Constants.HeaderConstants.X_MS_META + "-"
+                                + ENCRYPTION_DATA_KEY) == null) {
+                                return Mono.just(httpResponse);
+                            }
+                            tuple2.getT2().setAdjustedDownloadCount(
+                                Long.parseLong(responseHeaders.getValue(CONTENT_LENGTH)));
+
+                            /*
+                             * We expect padding only if we are at the end of a blob and it is not a multiple of the
+                             * encryption block size. Padding is only ever present in track 1.
+                             */
+                            boolean padding = tuple2.getT1().getEncryptionAgent().getProtocol()
+                                .equals(ENCRYPTION_PROTOCOL_V1)
+                                && (tuple2.getT2().toBlobRange().getOffset() + tuple2.getT2().toBlobRange().getCount()
+                                > (blobSize(responseHeaders) - ENCRYPTION_BLOCK_SIZE));
+
+                            Flux<ByteBuffer> plainTextData = this.decryptBlob(httpResponse.getBody(), tuple2.getT2(), padding,
+                                tuple2.getT1(), httpResponse.getRequest().getUrl().toString());
+
+                            return Mono.just(new DecryptedResponse(httpResponse, plainTextData));
+                        } else {
+                            return Mono.just(httpResponse);
+                        }
+                    }));
+            });
+        }
+    }
+
+    private BlobRequestConditions extractRequestConditionsFromRequest(HttpHeaders requestHeaders) {
+        return new BlobRequestConditions()
```

Comment:
for GA: we should introduce BlobRequestCondition(HttpHeaders) ctor.

Reply:
Added to list